### PR TITLE
adding a straight through estimator for relaxed categorical

### PR DESCRIPTION
**Pull request by karalets**
(builds on code from @karalets and @fritzo)

    An implementation of RelaxedOneHotCategorical with a straight-through gradient estimator.

    This distribution has the following properties:
    -- the samples returned by the `rsample` method are discrete/quantized
    -- the `log_prob` method returns the log probability of the relaxed/unquantized sample
       using the GumbelSoftmax distribution
    -- in the backward pass the gradient of the sample with respect to the parameters of the
       distribution uses the relaxed/unquantized sample

now has more legit tests. bernoulli can be done in another pr.

### CODE REVIEW

File: pyro/distributions/relaxed_straight_through.py [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+    In detail, it uses discrete samples for the forward step but the relaxed sample for the backwards gradient.
+    """
+    def __init__(self, temperature, probs=None, logits=None, validate_args=None):
+        super(RelaxedCategoricalStraightThrough, self).__init__(temperature=temperature, probs=probs, logits=logits, validate_args=validate_args)
+        self._unquantize = {}
+
+    def rsample(self, sample_shape=torch.Size()):
+        soft_sample = super(RelaxedCategoricalStraightThrough, self).rsample(sample_shape)
+        hard_sample = QuantizeCategorical.apply(soft_sample)
+        self._unquantize[id(hard_sample)] = soft_sample
```
Comment (martinjankowiak) [inline pyro/distributions/relaxed_straight_through.py]: why not use the tensor itself as key?

Comment (karalets) [reply pyro/distributions/relaxed_straight_through.py]: ok, the hard_sample is a onehot vector mapping of a dense vector.
However, multiple dense vectors can map to the same onehot vector.

As such, in order to make the mapping unique, we decided to use the id instead of the value.

Makes sense?

File: pyro/distributions/relaxed_straight_through.py [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+    """
+    def __init__(self, temperature, probs=None, logits=None, validate_args=None):
+        super(RelaxedOneHotCategoricalStraightThrough, self).__init__(temperature=temperature, probs=probs,
+                                                                      logits=logits, validate_args=validate_args)
+        self._unquantize = {}
+
+    def rsample(self, sample_shape=torch.Size()):
+        soft_sample = super(RelaxedOneHotCategoricalStraightThrough, self).rsample(sample_shape)
+        hard_sample = QuantizeCategorical.apply(soft_sample)
+        self._unquantize[id(hard_sample)] = soft_sample
```
Comment (martinjankowiak) [inline pyro/distributions/relaxed_straight_through.py]: would it be better to use `hard_sample.data_ptr()` instead of `id()`

File: pyro/distributions/relaxed_straight_through.py [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+        argmax = input.max(-1)[1]
+        result = input.new_zeros(input.shape)
+        # result[argmax] = 1
+        if argmax.dim() < result.dim():
+            argmax = argmax.unsqueeze(-1)
+        return result.scatter_(-1, argmax, 1)
+
+    @staticmethod
+    def backward(ctx, grad):
+        return grad
```
Comment (martinjankowiak) [inline pyro/distributions/relaxed_straight_through.py]: there was a `grad.clone()` here. i got rid of it. shouldn't be needed?

File: tests/distributions/test_relaxed_straight_through.py [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+        temperature = torch.tensor(0.5)
+        probs = torch.tensor(probs, requires_grad=True)
+        d = RelaxedOneHotCategoricalStraightThrough(temperature, probs=probs)
+        sample = d.rsample()
+        log_prob = d.log_prob(sample)
+        grad_probs = grad(log_prob.sum(), [probs])[0]
+        assert grad_probs.shape == probs.shape
+
+
+@pytest.mark.parametrize('temp', [0.2, 0.5, 1.0])
```
Comment (martinjankowiak) [inline tests/distributions/test_relaxed_straight_through.py]: this test basically just tests plumbing

File: tests/distributions/test_relaxed_straight_through.py [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+    dist_q = RelaxedOneHotCategoricalStraightThrough(temperature=temp, probs=q)
+    z = dist_q.rsample(sample_shape=(num_samples,))
+    actual = grad(dist_q.log_prob(z).sum(), [q])[0] / num_samples
+
+    assert_equal(expected, actual, prec=0.1,
+                 msg='bad grad for RelaxedOneHotCategoricalStraightThrough (expected {}, got {})'.
+                 format(expected, actual))
+
+
+def test_svi_usage():
```
Comment (martinjankowiak) [inline tests/distributions/test_relaxed_straight_through.py]: this test actually shows how the (biased) gradients can be used in svi

**Comment by martinjankowiak** [comment]
cc @karalets take a look should be ready to go

File: pyro/distributions/relaxed_straight_through.py (hunk 2 above)
Comment (fritzo) [reply pyro/distributions/relaxed_straight_through.py]: I'd omit the `id()` entirely since `Tensor` already hashes on id. `.data_ptr()` is too clever imo.

File: pyro/distributions/relaxed_straight_through.py (hunk 3 above)
Comment (fritzo) [reply pyro/distributions/relaxed_straight_through.py]: sounds reasonable, i do not understand the purpose of the `.clone()`

File: pyro/distributions/__init__.py [hunk 6]
```diff
@@ -33,6 +34,7 @@
     "Rejector",
     "TorchDistribution",
     "VonMises",
+    "RelaxedOneHotCategoricalStraightThrough"
```
Comment (fritzo) [inline pyro/distributions/__init__.py]: nit: make these alphabetical and add a trailing comma.

File: pyro/distributions/relaxed_straight_through.py [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+
+import torch
+
+from pyro.distributions.torch import RelaxedOneHotCategorical
+from pyro.distributions.util import copy_docs_from
+
+
+@copy_docs_from(RelaxedOneHotCategorical)
+class RelaxedOneHotCategoricalStraightThrough(RelaxedOneHotCategorical):
+    """
```
Comment (martinjankowiak) [inline pyro/distributions/relaxed_straight_through.py]: hope these docs come out looking the way they ought?

File: pyro/distributions/relaxed_straight_through.py (hunk 2 above)
Comment (fritzo) [reply pyro/distributions/relaxed_straight_through.py]: Actually this is just plain buggy. I believe a correct implementation could instead use a [WeakKeyDictionary](https://docs.python.org/2/library/weakref.html#weakref.WeakKeyDictionary).

File: tests/distributions/test_relaxed_straight_through.py [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+        d = RelaxedOneHotCategoricalStraightThrough(temperature, probs=probs)
+        sample = d.rsample()
+        log_prob = d.log_prob(sample)
+        grad_probs = grad(log_prob.sum(), [probs])[0]
+        assert grad_probs.shape == probs.shape
+
+
+@pytest.mark.parametrize('temp', [0.2, 0.5, 1.0])
+def test_entropy_grad(temp):
+    num_samples = 1000000
```
Comment (fritzo) [inline tests/distributions/test_relaxed_straight_through.py]: Why does 1e6 samples only lead to `prec=0.1`? Is this due to finite temperature effects?

Comment (martinjankowiak) [reply tests/distributions/test_relaxed_straight_through.py]: it's partially that. it's also that the scale isn't 1.0. some of the elements of `expected` are of size 7.6 or what not

File: pyro/distributions/relaxed_straight_through.py (hunk 2 above)
Comment (martinjankowiak) [reply pyro/distributions/relaxed_straight_through.py]: ok well it would be great if, once everything else in this PR is in order, you could push the needed changes. because i don't know what you have in mind

File: pyro/distributions/relaxed_straight_through.py (hunk 7 above)
Comment (jpchen) [reply pyro/distributions/relaxed_straight_through.py]: if only there existed something like `make docs` for you to see them...

File: pyro/distributions/relaxed_straight_through.py (hunk 2 above)
Comment (fritzo) [reply pyro/distributions/relaxed_straight_through.py]: done.

Comment (martinjankowiak) [reply pyro/distributions/relaxed_straight_through.py]: great thanks!

File: pyro/distributions/__init__.py (hunk 6 above)
Comment (martinjankowiak) [reply pyro/distributions/__init__.py]: done

**Comment by karalets** [comment]
lgtm after my questions in person, please merge
